improve retry backoff for web hook patching during conflict

ramaraochavali opened the pull request:
When we moved from Istio Queue to k8s Queue in https://github.com/istio/istio/pull/35392, we lost the ability to retry with exponential backoff on conflicts added in https://github.com/istio/istio/pull/35243. So when the problem described there comes, we still end up bringing down etcd.

This PR adds in-place retry with backoff on conflict.

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

ramaraochavali committed:
retry with exponential backoff when patch ends in conflict

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1601524705183928320) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1601524705183928320) | true | `/test unit-tests-arm64`
unit-tests_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests_istio/1601524705137790976) | true | `/test unit-tests`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1601524705183928320) | true | `/test unit-tests-arm64`
unit-tests_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests_istio/1601524705137790976) | true | `/test unit-tests`
release-test_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/release-test_istio/1601524705221677056) | true | `/test release-test`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1601524705183928320) | true | `/test unit-tests-arm64`
unit-tests_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests_istio/1601524705137790976) | true | `/test unit-tests`
release-test_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/release-test_istio/1601524705221677056) | true | `/test release-test`
integ-basic-arm64_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/integ-basic-arm64_istio/1601524705951485952) | true | `/test integ-basic-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1601524705183928320) | true | `/test unit-tests-arm64`
unit-tests_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests_istio/1601524705137790976) | true | `/test unit-tests`
release-test_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/release-test_istio/1601524705221677056) | true | `/test release-test`
integ-basic-arm64_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/integ-basic-arm64_istio/1601524705951485952) | true | `/test integ-basic-arm64`
integ-security-multicluster_istio | 3d1df8c08f272776fbfa8f895351ece6c72a50ca | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/integ-security-multicluster_istio/1601524711810928640) | true | `/test integ-security-multicluster`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali committed:
fix compile

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

howardjohn commented:
If we lost the proper retrying, maybe its time to add a test for it?

One note on this - if we have multiple items in the queue this will block all of them

ramaraochavali commented:
> If we lost the proper retrying, maybe its time to add a test for it?

+1 . I will add a test for it in this PR.

ramaraochavali committed:
add test

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali commented:
@howardjohn  I have added test. PTAL

> One note on this - if we have multiple items in the queue this will block all of them

This wont block all of them. It will try 5 times on conflict with a backoff and then returns with an error. It will be queued again later. This wont block all of them.

However, if multiple Istiod instances have inconsistent certs for longer duration, this would still put etcd under pressure.  Do you think if any thing can be done to slow that down while we keep retrying?  Should we delay queuing with additional backoff?

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio | adbb4786dd3ad42337944aabf982165e4b94da8c | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/lint_istio/1603717059747778560) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali committed:
fix lint

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali committed:
use slow fast rate limiter

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali commented:
@howardjohn changed implementation to use `NewItemFastSlowRateLimiter`  and retry forever (that used to be the behaviour earlier) so that we do initial retries fast, but if certs are in inconsistent state, we slow down the retries to protect etcd. Also added test to verify. PTAL if this makes sense

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | a65cdda76cb8628653ba7506e369b1b81dcf512b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1604044191908761600) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | a65cdda76cb8628653ba7506e369b1b81dcf512b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1604044191908761600) | true | `/test unit-tests-arm64`
unit-tests_istio | a65cdda76cb8628653ba7506e369b1b81dcf512b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests_istio/1604044191866818560) | true | `/test unit-tests`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | a65cdda76cb8628653ba7506e369b1b81dcf512b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests-arm64_istio/1604044191908761600) | true | `/test unit-tests-arm64`
unit-tests_istio | a65cdda76cb8628653ba7506e369b1b81dcf512b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/unit-tests_istio/1604044191866818560) | true | `/test unit-tests`
integ-pilot-istiodremote_istio | a65cdda76cb8628653ba7506e369b1b81dcf512b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/integ-pilot-istiodremote_istio/1604044196048539648) | true | `/test integ-pilot-istiodremote`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented:
I think to fix the conflict, the ideal way is to use leader election

ramaraochavali commented:
> I think to fix the conflict, the ideal way is to use leader election

I think there were concerns about leader election - we need to agree on a solution there. However this is to fix the regression that we introduced when we moved away from custom queue to controller queue, we lost the backoff capability.

ramaraochavali committed:
fix race

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-security-multicluster_istio | 6a84756bb7dfbb9fed307e5443bc32f783ab79eb | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/42387/integ-security-multicluster_istio/1604692588479123456) | true | `/test integ-security-multicluster`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-security-multicluster

hzxuzhonghu commented on pkg/webhooks/webhookpatch.go:
```diff
@@ -84,6 +84,17 @@ func NewWebhookCertPatcher(
+		// state for longer duration. Slowdown the retries, so that we do not overload kube api server and etcd.
+		controllers.WithRateLimiter(workqueue.NewItemFastSlowRateLimiter(100*time.Millisecond, 1*time.Minute, 5)),
+		// Webhook patching has to be retried forever. But the retries would be rate limited.
+		controllers.RetryForever())
```
Why not pass a maxint

ramaraochavali replied:
I can use that. This is to just make this clear that it is retried for ever.

hzxuzhonghu commented:
How does it recover if the retry interval is very long after a while

ramaraochavali commented:
> How does it recover if the retry interval is very long after a while

Retry interval is always 1m (it is not exponential backoff) for the same reason.

ramaraochavali committed:
use maxint

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>
